### ISSUE TITLE
Misleading error message with Trait and unimplemented std::marker::Sized

### ISSUE THREAD

**Issue by Weasy666** [rust-lang/rust#65734]
When doing something like this

``` rust
use std::path::{Path, PathBuf};

pub trait Control {
    type Output;
    fn open<P: AsRef<Path>>(path: P) -> Result<Self::Output, String>;
}

#[derive(Debug)]
pub struct Foo {
    path: Path,
}

impl Control for Foo {
    type Output = Self;

    fn open<P: AsRef<Path>>(path: P) -> Result<Self, String> {
        Ok(Foo {
            path: path.as_ref().into(),
        })
    }
}

fn main() {
    let path = Path::new("/test");
    let c = Foo::open(path);
    print!("{:?}", c);
}
```
> [Playground Link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7da2ecc159bd0be36fa7190c2b5a7554)

you get an error like this
```
error[E0277]: the trait bound `std::path::Path: std::convert::AsRef<&std::path::Path>` is not satisfied
  --> src/main.rs:26:23
   |
6  |     fn open<P: AsRef<&'a Path>>(path: P) -> Result<Self::Output, String>;
   |     --------------------------------------------------------------------- required by `Control::open`
...
26 |     let c = Foo::open(path);
   |                       ^^^^ the trait `std::convert::AsRef<&std::path::Path>` is not implemented for `std::path::Path`
   |
   = help: the following implementations were found:
             <std::path::Path as std::convert::AsRef<std::ffi::OsStr>>
             <std::path::Path as std::convert::AsRef<std::path::Path>>
   = note: required because of the requirements on the impl of `std::convert::AsRef<&std::path::Path>` for `&std::path::Path`

```

---
I would expect an error like the one from this code
``` rust
use std::path::{Path, PathBuf};

pub trait Control<'a> {
    type Output;
    fn open<P: 'a + AsRef<Path>>(path: P) -> Result<Self::Output, String>;
}

pub struct Foo<'a> {
    path: &'a Path,
}

impl<'a> Control<'a> for Foo<'a> {
    type Output = Self;

    fn open<P: 'a + AsRef<Path>>(path: P) -> Result<Self, String> {
        Ok(Foo {
            path: path.as_ref(),
        })
    }
}

fn main() {
    let path  = PathBuf::from("/test");
    let c = Foo::open(path  );
}
```
> [Playground Link](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=49205816abdac866ddf01a7bd33499d5)

```
error[E0515]: cannot return value referencing function parameter `path`
  --> src/main.rs:16:9
   |
16 | /         Ok(Foo {
17 | |             path: path.as_ref(),
   | |                   ---- `path` is borrowed here
18 | |         })
   | |__________^ returns a value referencing data owned by the current function
```

**Comment by estebank** [comment]
Current output:

```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> src/main.rs:17:41
   |
17 |     fn open<P: AsRef<Path>>(path: P) -> Result<Self, String> {
   |                                         ^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: within `Foo`, the trait `Sized` is not implemented for `[u8]`
note: required because it appears within the type `Foo`
  --> src/main.rs:10:12
   |
10 | pub struct Foo {
   |            ^^^
note: required by a bound in `Result`
  --> /rustc/f3126500f25114ba4e0ac3e76694dd45a22de56d/library/core/src/result.rs:503:1
```
